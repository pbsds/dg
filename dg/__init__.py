import os
import sys
import types

try:
    PY_TAG     = sys.implementation.cache_tag or 'unknown'
    PY_VERSION = sys.hexversion
    BUNDLE_DIR = os.path.join(__path__[0], 'bundle')
except AttributeError:
    raise ImportError('Python >= 3.4 required')

def load():
    from marshal import load

    try:
        with open(os.path.join(BUNDLE_DIR, PY_TAG + '.dgbundle'), 'rb') as fd:
            for code in load(fd):
                eval(code)
    except FileNotFoundError:
        try:
            with open(os.path.join(BUNDLE_DIR, PY_TAG + '.dgbundle.py')) as fd:
                for code in eval(fd.read(), {'C': types.CodeType}):
                    eval(code)
        except FileNotFoundError:
            raise ImportError('python implementation {!r} not supported'.format(PY_TAG))

load()
