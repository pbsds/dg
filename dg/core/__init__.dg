import '/os'
import '/types'
import '/marshal'
import '/builtins'


#: Recompile this directory.
#:
#: old     -- the current version of the bundle.
#: tag     -- `sys.implementation.cache_tag` of the target.
#: version -- `sys.hexversion` of the target.
#:
#: build :: [CodeType] str int -> [CodeType]
#:
build = old tag version bundle_tag: None ->
  ns = dict __name__: 'dg' PY_TAG: tag PY_VERSION: version __builtins__: builtins.__dict__
  for c in old => eval c ns
  bundle_tag is None or
    yield $ (ns !! 'compile') ('BUNDLE_TAG = {!r}'.format bundle_tag)
  for d in __path__ =>
    for f in sorted (os.listdir d) => if not $ f.startswith '_' =>
      with fd = open (os.path.join d f) 'r' =>
        yield $ (ns !! 'compilefd') fd ('<core:' + f + '>')


#: Load an existing bundle.
#:
#: tag     -- `sys.implementation.cache_tag` of the target interpreter.
#: dirname -- path to the directory with bundles.
#:
#: load :: str str -> [CodeType]
#:
load = tag dirname ->
  except
    err => with fd = open (os.path.join dirname (tag + '.dgbundle')) 'rb' => marshal.load fd
    err :: FileNotFoundError => with fd = open (os.path.join dirname (tag + '.dgbundle.py')) => eval fd.read! {('C', types.CodeType)}


#: Write the compiled code bundle as returned by `rebuild`.
#:
#: tag     -- `sys.implementation.cache_tag` of the target interpreter.
#: dirname -- path to the directory with bundles.
#:
#: save :: [CodeType] str str -> a
#:
save = code tag dirname ->
  data = marshal.dumps code
  os.makedirs dirname exist_ok: True
  # Note that `marshal.dump` may raise an exception mid-file,
  # leaving the bundle in a corrupted state. That generally means there
  # is an error in the compiler, so opening the file before finishing the
  # encoding process is undesirable.
  with fd = open (os.path.join dirname (tag + '.dgbundle')) 'wb' => fd.write data


_repr = x -> if
  x :: types.CodeType =>
    fields  = ['argcount', 'kwonlyargcount', 'nlocals', 'stacksize', 'flags', 'code', 'consts', 'names']
    fields += ['varnames', 'filename', 'name', 'firstlineno', 'lnotab', 'freevars', 'cellvars']
    'C({})'.format $ ','.join $ map (f -> _repr $ getattr x $ 'co_' + f) fields
  x :: list  => '[{}]'.format $ ''.join $ map (i -> _repr i + ',') x
  x :: tuple => '({})'.format $ ''.join $ map (i -> _repr i + ',') x
  otherwise => repr x


#: Write the compiled bundle without using `marshal`.
#:
#: save_compat :: [CodeType] str str -> a
#:
save_compat = code tag dirname ->
  data = _repr code
  os.makedirs dirname exist_ok: True
  with fd = open (os.path.join dirname (tag + '.dgbundle.py')) 'w' => fd.write data
