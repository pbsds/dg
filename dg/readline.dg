import '/os'
import '/re'
import '/time'
import '/math'
import '/fcntl'
import '/struct'
import '/select'
import '/termios'
import '/contextlib'
import '/collections'


#: ANSI character DFA. Matches either a single character or a complete control sequence.
#:
#: Sequence :: Re
#:
Sequence = re.compile r'''(?sx)\033(?:
  \[.*?[@-~]     # ANSI CSI: standard escape sequences.
 |\]\d+;.*?\007  # ANSI OSC: e.g. sets the title of an XTerm-like VT.
 |[NO].          # ANSI SS2/SS3: alternate character sets.
)|(.)'''


#: A sequence of characters considered by the real GNU readline to be zero-length.
ZeroLengthSequence = re.compile r'''\001.*?\002'''


Key = subclass object where
  DEL_ASCII = '\x7f'     # This is actually used as Backspace. Probably.
  DEL_ANSI  = '\033[3~'  # *This* is Delete.
  INSERT    = '\033[2~'
  PAGE_UP   = '\033[5~'
  PAGE_DOWN = '\033[6~'
  UP        = '\033[A'
  DOWN      = '\033[B'
  RIGHT     = '\033[C'
  LEFT      = '\033[D'
  HOME      = '\033[H'
  END       = '\033[F'
  HOME_ALT  = '\033OH'  # Non-standard sequences?
  END_ALT   = '\033OF'

  # Control sequence = ASCII name [= known name] = value
  CONTROL_A = SOH = HOME_CTRL  = '\x01'
  CONTROL_B = STX = LEFT_CTRL  = '\x02'
  CONTROL_C = ETX = SIGINT     = '\x03'  # only usable via `enable_jobs = False`
  CONTROL_D = EOF              = '\x04'
  CONTROL_E = ENQ = END_CTRL   = '\x05'
  CONTROL_F = ACK = RIGHT_CTRL = '\x06'
  CONTROL_G = BEL = ALARM      = '\x07'
  CONTROL_H = BS  = DEL_CTRL   = '\x08'
  CONTROL_I = TAB              = '\x09'
  CONTROL_J = LF  = ENTER      = '\x0a'
  CONTROL_K = VT  = CUT_END    = '\x0b'
  CONTROL_L = FF  = NEWPAGE    = '\x0c'
  # Control+M = '\r' is unusable due to automatic line break conversion.
  CONTROL_N = SO  = DOWN_CTRL  = '\x0e'
  CONTROL_O = SI  = CUT_START  = '\x0f'
  CONTROL_P = DLE = UP_CTRL    = '\x10'
  CONTROL_Q = DC1 = XON        = '\x11'  # only usable via `enable_flowctl = False`
  CONTROL_R = DC2 = SEARCH     = '\x12'
  CONTROL_S = DC3 = XOFF       = '\x13'  # only usable via `enable_flowctl = False`
  CONTROL_T = DC4              = '\x14'
  CONTROL_U = NAK              = '\x15'
  CONTROL_V = SYN              = '\x16'
  CONTROL_W = ETB              = '\x17'
  CONTROL_X = CAN = CANCEL     = '\x18'
  CONTROL_Y = EM  = PASTE      = '\x19'
  CONTROL_Z = SUB = SUSPEND    = '\x1a'  # only usable via `enable_jobs = False`


Color = subclass object where
  FG = subclass object where
    black   = '\033[30m{}\033[39m'.format
    red     = '\033[31m{}\033[39m'.format
    green   = '\033[32m{}\033[39m'.format
    yellow  = '\033[33m{}\033[39m'.format
    blue    = '\033[34m{}\033[39m'.format
    magenta = '\033[35m{}\033[39m'.format
    cyan    = '\033[36m{}\033[39m'.format
    white   = '\033[37m{}\033[39m'.format

  BG = subclass object where
    black   = '\033[40m{}\033[49m'.format
    red     = '\033[41m{}\033[49m'.format
    green   = '\033[42m{}\033[49m'.format
    yellow  = '\033[43m{}\033[49m'.format
    blue    = '\033[44m{}\033[49m'.format
    magenta = '\033[45m{}\033[49m'.format
    cyan    = '\033[46m{}\033[49m'.format
    white   = '\033[47m{}\033[49m'.format


Teletype = subclass object where
  #: Various TTY-related functions.
  #:
  #: input  :: TextIOBase -- 'r'-mode TTY.
  #: output :: TextIOBase -- 'w'-mode TTY.
  #:
  __init__ = @input @output @name: None ~>
    not @input .isatty! => raise $ IOError 'input: not a tty'
    not @output.isatty! => raise $ IOError 'output: not a tty'
    None

  read  = ~> @input.read
  write = ~> @output.write
  flush = ~> @output.flush!

  #: Return true iff input is a single non-control character.
  #:
  #: printable :: str -> bool
  #:
  printable = classmethod $ data ~> if
    len data == 0 => True
    len data == 1 => 32 <= ord data < 127 or 160 <= ord data

  #: Enter the raw mode of a TTY: disable line buffering, echo, etc.
  #:
  #:   * `jobctl`: whether to enable Ctrl+C (SIGINT) and Ctrl+Z (SIGSTOP).
  #:   * `flowctl`: whether to enable Ctrl+S (XOFF) and Ctrl+Q (XON).
  #:
  #: where "enable" means the default behavior. That is, pass `False` to bind your
  #: own actions on these combos. Otherwise, they won't even reach the application.
  #:
  #: raw :: bool bool -> context NoneType
  #:
  raw = contextlib.contextmanager $ jobctl flowctl ~>
    old = termios.tcgetattr @input
    new = termios.tcgetattr @input
    new !! 0 &= ~(termios.ISTRIP | termios.IXON)
    new !! 3 &= ~(termios.ECHO   | termios.ICANON | termios.IEXTEN | termios.ISIG)
    new !! 6 !! termios.VMIN  = 0  # This makes `read` non-blocking.
    new !! 6 !! termios.VTIME = 0
    if flowctl => new !! 0 |= termios.IXON
    if jobctl  => new !! 3 |= termios.ISIG
    except _       => yield $ termios.tcsetattr @input termios.TCSADRAIN new
           finally =>         termios.tcsetattr @input termios.TCSAFLUSH old

  #: Find out how much stuff this TTY can display at once.
  #:
  #: size :: (int, int) -- (height, width) in characters
  #:
  size = ~> struct.unpack 'hh' $ fcntl.ioctl @output termios.TIOCGWINSZ '1234'

  #: Read at most a line from a TTY (blocking until at least a character is available.)
  #:
  #: buffer :: float -> iter str
  #:
  buffer = timeout ~>
    x = @input.readline!
    x == '' =>
      has_data = fst $ if timeout   => select.select [@input] [] [] timeout
                          otherwise => select.select [@input] [] []
      x = @input.readline!
      x == '' and has_data => raise $ IOError 'input channel closed'
    map (x -> x.group!) $ Sequence.finditer x

  #: Move the caret `y` characters down and `x` characters to the right.
  #:
  #: relmove :: int int -> a
  #:
  relmove = y x ~> @write $ (+)
    if y > 0     => '\033[{}B'.format y
       y < 0     => '\033[{}A'.format (-y)
       otherwise => ''  # If `y` is 0, terminals might assume 1.
    if x > 0     => '\033[{}C'.format x
       x < 0     => '\033[{}D'.format (-x)
       otherwise => ''

  #: Erase everything from the cursor to the end of the screen.
  #:
  #: erase :: a
  #:
  erase = ~> @write '\033[J'


ReadlineState = subclass object where
  #: A readline input state.
  #:
  #: Defines stuff such as the current buffer contents, caret position, etc.
  #:
  __init__ = @prompt @buffer @position @cache ~> None

  #: Move the caret to the right.
  #:
  #: right :: int -> ReadlineState
  #:
  right = i ~> ReadlineState @prompt @buffer (@position + i `min` len @buffer) @cache

  #: Move the caret to the left.
  #:
  #: left :: int -> ReadlineState
  #:
  left = i ~> ReadlineState @prompt @buffer (@position - i `max` 0) @cache

  #: Insert a string into the buffer, move the caret if necessary.
  #:
  #: insert :: int str -> ReadlineState
  #:
  insert = i d ~> if
    i < 0     => @insert 0 $ d !! slice (-i) None
    otherwise => ReadlineState @prompt
      (+)
        @buffer !! slice None i + d
        @buffer !! slice i None
      @position + len d * (i <= @position)
      @cache

  #: Erase the characters at position `i`, move the caret accordingly.
  #:
  #: erase :: int (Optional int) -> ReadlineState
  #:
  erase = i length: 1 ~> if
    i < 0     => @erase 0 $ length + i
    otherwise => ReadlineState @prompt
      (+)
        @buffer !! slice None i
        @buffer !! slice (i + length) None
      @position - (length `min` @position - i `max` 0)
      @cache


Readline = subclass object where
  __init__ = *: args **: kwargs ~>
    @tty = Teletype *: args **: kwargs
    # In case subclasses decide to override keybindings, make sure
    # they don't add them to the global mapping.
    @bindings = @bindings.copy!
    @bindings !! Key.CONTROL_L = state ->
      state = @erase state
      @tty.write '\033[H\033[J'
      state
    None

  #: A mapping of keys to state modificators. Note that a key may be
  #: either a character or an ANSI sequence.
  #:
  #: bindings :: dict str (ReadlineState -> ReadlineState)
  #:
  bindings = dict'
    Key.ENTER,      state -> raise $ StopIteration state.buffer
    Key.RIGHT,      state -> state.right 1
    Key.RIGHT_CTRL, state -> state.right 1
    Key.LEFT,       state -> state.left  1
    Key.LEFT_CTRL,  state -> state.left  1
    Key.END,        state -> state.right $ len state.buffer
    Key.END_ALT,    state -> state.right $ len state.buffer
    Key.END_CTRL,   state -> state.right $ len state.buffer
    Key.HOME,       state -> state.left    state.position
    Key.HOME_ALT,   state -> state.left    state.position
    Key.HOME_CTRL,  state -> state.left    state.position
    Key.DEL_ANSI,   state -> state.erase   state.position
    Key.DEL_ASCII,  state -> state.erase $ state.position - 1
    Key.DEL_CTRL,   state -> state.erase $ state.position - 1
    Key.EOF,        state -> if
      state.buffer => state.erase state.position
      otherwise    => raise EOFError

  #: Whether to enable ^C and ^Z. Disallows the use of `Key.SIGINT` and `Key.SUSPEND`.
  #:
  #: enable_jobs :: bool
  #:
  enable_jobs = True

  #: Whether to enable XON/XOFF, whatever that means. Disables `Key.XON` and `Key.XOFF`.
  #:
  #: enable_flowctl :: bool
  #:
  enable_flowctl = False

  #: Default line preprocessor. Does nothing.
  #:
  #: This method is called whenever an updated display is required.
  #: Its only argument is the buffer, and the return value should
  #: be an ANSI-escaped string to display. This could be used,
  #: for example, to implement syntax highlighting.
  #:
  #: NOTE: this method must NOT change the apparent length of a string.
  #:   Otherwise, the caret will appear in the wrong position.
  #:
  #: preprocess :: str -> str
  #:
  preprocess = ln ~> ln

  #: Read a line from a TTY while displaying a preview on another TTY.
  #:
  #: If the prompt is a function, it will be called with the current
  #: `ReadlineState` whenever the display is updated.
  #:
  #: input :: (Either str (ReadlineState -> str)) (Optional float) -> str
  #:
  input = prompt timeout: 0 buffer: '' position: 0 ~>
    with @tty.raw @enable_jobs @enable_flowctl => except
      err =>
        start = time.time!
        state = ReadlineState prompt buffer position {}
        while True =>
          passed = min timeout $ time.time! - start
          passed >= timeout > 0 => raise $ BlockingIOError state
          @update state

          for c in @tty.buffer (timeout - passed) => if
            @bindings.get  c => state = (@bindings !! c) state
            @tty.printable c => state = state.insert state.position c
            otherwise        => @tty.write Key.ALARM
      err :: StopIteration => err.value
      finally => @finalize state

  #: Write a preview of a line buffer onto a TTY. Writing stuff to output afterwards
  #: will mess with the prompt; call `erase` or `finalize` first.
  #:
  #: update :: ReadlineState -> ReadlineState
  #:
  update = state ~>
    @erase state
    prefix = if
      state.prompt :: collections.Callable => state.prompt state
      otherwise => str state.prompt
    prelen = len $ ZeroLengthSequence.sub '' prefix
    wh, ww = @tty.size
    h, w = prelen + len state.buffer `divmod` ww
    y, x = prelen + state.position   `divmod` ww
    @tty.write $ if
      h < wh    => prefix + @preprocess state.buffer
      otherwise => ''.join
        where
          # Not enough lines to display everything, let's try another approach.
          # We'll only show the `ww - 1` characters closest to the cursor.
          offset = prelen + state.position
          start = max (offset - (ww - 1 >> 1)) 0
          end   = min (offset + (ww - 1 >> 1)) (prelen + len state.buffer)
          # And again, to ensure that we show `ww - 1` characters
          # even if the cursor is at either end of the buffer.
          end   = min (max end $ start + ww - 1) (prelen + len state.buffer)
          start = max (min start $ end - ww + 1) 0
          y, x := 0, offset - start
          h, w := 0, end    - start
          c = 0
          for m in Sequence.finditer (prefix + @preprocess state.buffer) => if
            m.group 1 => start < (c += 1) <= end => yield $ m.group 0
            # Include all ANSI escapes in the output to ensure that
            # all resets work as intended.
            otherwise => yield $ m.group 0
    # Some TTYs don't create a new line if the output fits the window exactly;
    # the problem is, we still have to show the cursor, but there's no space.
    # So we need to create a new line to put it on manually.
    h != 0 and w == 0 => @tty.write '\n\r'
    @tty.relmove (y - h) (x - w)
    @tty.flush
    # `erase` will use that info to locate the start of the buffer.
    # FIXME this value may become invalid if the terminal rewraps on resize.
    state.cache !! 'offset' = y, x
    state

  #: Completely erase the prompt and the input buffer from the screen.
  #:
  #: erase :: ReadlineState -> ReadlineState
  #:
  erase = state ~>
    y, x = state.cache.pop 'offset' (0, 0)
    @tty.relmove (-y) (-x)
    @tty.erase
    state

  #: Make sure the prompt does not interfere with further output to the TTY.
  #:
  #: finalize :: ReadlineState -> ReadlineState
  #:
  finalize = state ~>
    @erase state
    prefix = if
      state.prompt :: collections.Callable => state.prompt state
      otherwise => str state.prompt
    @tty.write $ prefix + @preprocess state.buffer + '\n\r'
    @tty.flush
    state


History = subclass Readline where
  #: A mix-in that enables history navigation with up and down arrows.
  #:
  #: history :: [str] -- a list of previously entered lines
  #:
  __init__ = *: args **: kwargs ~>
    (super History self).__init__ *: args **: kwargs
    @history = list!
    @bindings.update $ dict'
      Key.UP,        @history_prev
      Key.UP_CTRL,   @history_prev
      Key.DOWN,      @history_next
      Key.DOWN_CTRL, @history_next
    None

  #: Switch to the previous history entry, if any. The contents of the buffer are lost.
  #:
  #: history_prev :: ReadlineState -> ReadlineState
  #:
  history_prev = state ~>
    i = state.cache.get 'history' -1
    if i <= -len @history => @tty.write Key.ALARM
       otherwise =>
         state = state.erase  0 $ len state.buffer
         state = state.insert 0 $ @history !! (i - 1)
         state.cache !! 'history' = i - 1
    state

  #: Switch to the next history entry, if any. The contents of the buffer are lost.
  #:
  #: history_next :: ReadlineState -> ReadlineState
  #:
  history_next = state ~>
    i = state.cache.get 'history' -1
    if i >= -1   => @tty.write Key.ALARM
       otherwise =>
         state = state.erase  0 $ len state.buffer
         state = state.insert 0 $ @history !! (i + 1)
         state.cache !! 'history' = i + 1
    state

  input = *: args **: kwargs ~> result where
    @history.append ''
    @history !! -1 = result = (super History self).input *: args **: kwargs
    @history !! -1 or @history !!~ -1


Completer = subclass Readline where
  completion_keys  = [Key.TAB]
  completion_regex = re.compile r'\w+'

  #: A mix-in that enables basic completion.
  #:
  #: completion_keys  :: [str] -- the keys on which to trigger completion.
  #: completion_regex :: Re  -- a regex that matches all completable words.
  #:
  __init__ = *: args **: kwargs ~>
    (super Completer self).__init__ *: args **: kwargs
    for k in @completion_keys => @bindings !! k = @complete_event
    None

  #: Handle a completion event. Default implementation calls `complete_buffer`
  #: and either substitutes the incomplete word for a result or outputs the list
  #: with `display_completions`.
  #:
  #: complete_event :: ReadlineState -> ReadlineState
  #:
  complete_event = state ~>
    @tty.write Key.ALARM
    start, length, cmps = @complete_buffer state.buffer state.position
    if len cmps == 1 => state = state.erase start length |>.insert start (head cmps)
       len cmps != 0 =>
         prefix = ''.join $ map fst $ takewhile (cs -> len (set cs) == 1) $ zip *: cmps
         # e.g. if all possible completions for `a` are `abc1`, `abc2`, and `abc3`,
         #      this will automatically insert `abc` after displaying the list.
         state = @finalize state |>.erase start length |>.insert start prefix
         @display_completions cmps
    state

  #: Find the word to complete and get the list of possible completions.
  #: Returns a (start index, length, completions) tuple.
  #:
  #: complete_buffer :: str int -> (int, int, [str])
  #:
  complete_buffer = buf at ~> except
    err => m = next $ filter (m -> m.start! <= at <= m.end!) $ @completion_regex.finditer buf
    err is None => m.start!, m.end! - m.start!, @complete_word m.group!
    err :: StopIteration => 0, 0, []

  #: Fetch the list of possible completions for a word.
  #:
  #: complete_word :: str -> [str]
  #:
  complete_word = word ~> ["Completer.complete_word", "not", "implemented"]

  #: Format and display a list of multiple possible completions.
  #:
  #: display_completions :: [str] -> a
  #:
  display_completions = cmps ~>
    maxlen = max $ map len cmps
    width  = @tty.size !! 1 // (maxlen + 1)
    height = math.ceil $ len cmps / width
    for y in range height =>
      for x in range y (len cmps `min` y + width * height) height =>
        @tty.write ' '
        @tty.write $ (cmps !! x).ljust maxlen
      @tty.write '\n'
