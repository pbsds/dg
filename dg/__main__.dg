import '/os'
import '/sys'
import '/runpy'
import '/importlib'

import '/dg'
import '/dg/core'
import '/dg/repl'

BUNDLE_TAG = 3177897669

USAGE = '''
python3 [options] -m dg -h
python3 [options] -m dg -b [tag [hexversion]]
python3 [options] -m dg [-c <command> | -m <module> | <file>] ...

VM options that make sense when used with dg:

  -B           don't save bytecode cache (i.e. `.pyc` and `.pyo` files);
  -E           ignore environment variables
  -s           don't add user site directory to sys.path
  -S           don't `import site` on startup
  -u           don't buffer standard output
  -v           trace import statements
  -X           implementation-specific options

Arguments:

  -h           show this message and exit
  -b           rebootstrap the compiler
    tag        cache tag of the target interpreter (e.g. `cpython-35`)
    hexversion version of the target interpreter as 8 hex digits (e.g. `030500A0`)
  -c command   run a single command, then exit
  -m module    run a module (or a package's `__main__`) as a script
  file         run a script
  ...          additional arguments accessible through `sys.argv`

Environment variables:

  PYTHONSTARTUP     a Python file executed on interactive startup
  PYTHONSTARTUPMOD  a name of the module to use instead of PYTHONSTARTUP
  PYTHONPATH        a `:`-separated list of directories to search for modules in
  PYTHONPREFIX      override the `--prefix` option of `configure`

...that have the same effect as options:

  PYTHONDONTWRITEBYTECODE    -B
  PYTHONNOUSERSITE           -s
  PYTHONUNBUFFERED           -u
  PYTHONVERBOSE              -v
'''

_arg = -> if
  len sys.argv > 1 => sys.argv.pop 1
  otherwise        => exit $ 'error: {} requires an argument'.format mode

sys.argv.pop 0  # `-m`
sys.argv = mode, *_ = sys.argv or ['-',]

rebuild = tag version verbose ->
  this = core.load dg.PY_TAG dg.BUNDLE_DIR
  this = list $ core.build this tag version BUNDLE_TAG
  verbose => print '[INFO] build 1 (bootstrap) ok'
  if
    tag != dg.PY_TAG =>
      verbose => print '[WARN] building for a different VM; self-testing impossible.'
      core.save_compat this tag dg.BUNDLE_DIR
    otherwise =>
      this = list $ core.build this tag version BUNDLE_TAG
      this = list $ core.build this tag version BUNDLE_TAG
      verbose => print '[INFO] builds 2 & 3 (test) ok'
      core.save this tag dg.BUNDLE_DIR

if mode != '-b' and getattr dg 'BUNDLE_TAG' None != BUNDLE_TAG =>
   except err => rebuild dg.PY_TAG dg.PY_VERSION verbose: False
          err :: Exception => print '[WARN] dg bundle outdated, and rebuild failed; try python -m dg -b' file: sys.stderr

if mode == '-h' => print USAGE
   mode == '-b' =>
     tag, version = if
       len sys.argv > 2 => sys.argv !! 1, int (sys.argv !! 2) 16
       len sys.argv > 1 => sys.argv !! 1, dg.PY_VERSION
       otherwise        => dg.PY_TAG,     dg.PY_VERSION
     rebuild tag version verbose: True
   mode == '-m' => runpy.run_module _arg! (dict __builtins__: dg.BUILTINS) '__main__' True
   mode == '-c' => repl.runcode $ dg.compile _arg! '<command>'
   mode != '-'  => repl.runfile $ open encoding: 'utf-8' $ head sys.argv
   otherwise    => repl.runfile sys.stdin where if
     sys.flags.ignore_environment =>  # `python -E`
     (fn = os.environ.get 'PYTHONSTARTUPMOD') => importlib.import_module fn
     (fn = os.environ.get 'PYTHONSTARTUP')    => runpy.run_path fn {} '__main__'
