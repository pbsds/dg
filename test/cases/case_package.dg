print 1 sep:'\n2\t' *: (sorted$ globals!.keys!)
print 3 sep:'\n3\t' *: (sorted$ locals!.keys!)
print 4 sep:'\n5\t' *: (sorted$ __builtins__.keys!)
print 6 sep:'\n7\t' *: (sorted$ (import '/builtins' pure).__dict__.keys!)
