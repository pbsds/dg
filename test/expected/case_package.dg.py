import dg; __builtins__.update(dg.BUILTINS)
print(1, *sorted(globals().keys()), sep='\n2\t')
print(3, *sorted(locals().keys()), sep='\n3\t')
print(4, *sorted(__builtins__.keys()), sep='\n5\t')
print(6, *sorted(__builtins__['!import'](__builtins__['!importname'](
    '/builtins'), __builtins__['__package__'], False).__dict__.keys()), sep
    ='\n7\t')
