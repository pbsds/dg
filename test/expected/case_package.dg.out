1
2	__builtins__
2	__doc__
2	__file__
2	__loader__
2	__name__
2	__package__
2	__spec__
3
3	__builtins__
3	__doc__
3	__file__
3	__loader__
3	__name__
3	__package__
3	__spec__
4
5	!
5	!!
5	!!~
5	!=
5	!BaseException
5	!chain
5	!dict
5	!import
5	!importname
5	!keyword_union
5	!locals
5	!subclass
5	!super
5	$
5	%
5	&
5	*
5	**
5	+
5	+:
5	,
5	-
5	->
5	->>
5	..
5	...
5	/
5	//
5	:+
5	::
5	<
5	<-
5	<<
5	<=
5	==
5	>
5	>=
5	>>
5	ArithmeticError
5	AssertionError
5	AttributeError
5	BaseException
5	BlockingIOError
5	BrokenPipeError
5	BufferError
5	BytesWarning
5	ChildProcessError
5	ConnectionAbortedError
5	ConnectionError
5	ConnectionRefusedError
5	ConnectionResetError
5	DeprecationWarning
5	EOFError
5	Ellipsis
5	EnvironmentError
5	Exception
5	False
5	FileExistsError
5	FileNotFoundError
5	FloatingPointError
5	FutureWarning
5	GeneratorExit
5	IOError
5	ImportError
5	ImportWarning
5	IndentationError
5	IndexError
5	InterruptedError
5	IsADirectoryError
5	KeyError
5	KeyboardInterrupt
5	LookupError
5	MemoryError
5	ModuleNotFoundError
5	NameError
5	None
5	NotADirectoryError
5	NotImplemented
5	NotImplementedError
5	OSError
5	OverflowError
5	PendingDeprecationWarning
5	PermissionError
5	ProcessLookupError
5	RecursionError
5	ReferenceError
5	ResourceWarning
5	RuntimeError
5	RuntimeWarning
5	StopAsyncIteration
5	StopIteration
5	SyntaxError
5	SyntaxWarning
5	SystemError
5	SystemExit
5	TabError
5	TimeoutError
5	True
5	TypeError
5	UnboundLocalError
5	UnicodeDecodeError
5	UnicodeEncodeError
5	UnicodeError
5	UnicodeTranslateError
5	UnicodeWarning
5	UserWarning
5	ValueError
5	Warning
5	ZeroDivisionError
5	^
5	__build_class__
5	__debug__
5	__doc__
5	__import__
5	__loader__
5	__name__
5	__package__
5	__spec__
5	abs
5	all
5	and
5	any
5	ascii
5	bin
5	bind
5	bool
5	breakpoint
5	bytearray
5	bytes
5	callable
5	chain
5	chr
5	classmethod
5	compile
5	complex
5	copyright
5	credits
5	delattr
5	dict
5	dict'
5	dir
5	divmod
5	drop
5	dropwhile
5	enumerate
5	eval
5	exec
5	exhaust
5	exit
5	filter
5	finally
5	flip
5	float
5	foldl
5	foldl1
5	format
5	frozenset
5	fst
5	getattr
5	globals
5	hasattr
5	hash
5	head
5	help
5	hex
5	id
5	in
5	init
5	input
5	int
5	is
5	isinstance
5	issubclass
5	iter
5	iterate
5	last
5	len
5	license
5	list
5	list'
5	locals
5	map
5	max
5	memoryview
5	min
5	next
5	not
5	object
5	oct
5	open
5	or
5	ord
5	otherwise
5	pow
5	print
5	property
5	quit
5	raise
5	range
5	repr
5	reversed
5	round
5	scanl
5	scanl1
5	set
5	set'
5	setattr
5	slice
5	snd
5	sorted
5	staticmethod
5	str
5	sum
5	super
5	tail
5	take
5	takewhile
5	tuple
5	tuple'
5	type
5	vars
5	zip
5	|
5	~
5	~>
5	~>>
6
7	ArithmeticError
7	AssertionError
7	AttributeError
7	BaseException
7	BlockingIOError
7	BrokenPipeError
7	BufferError
7	BytesWarning
7	ChildProcessError
7	ConnectionAbortedError
7	ConnectionError
7	ConnectionRefusedError
7	ConnectionResetError
7	DeprecationWarning
7	EOFError
7	Ellipsis
7	EnvironmentError
7	Exception
7	False
7	FileExistsError
7	FileNotFoundError
7	FloatingPointError
7	FutureWarning
7	GeneratorExit
7	IOError
7	ImportError
7	ImportWarning
7	IndentationError
7	IndexError
7	InterruptedError
7	IsADirectoryError
7	KeyError
7	KeyboardInterrupt
7	LookupError
7	MemoryError
7	ModuleNotFoundError
7	NameError
7	None
7	NotADirectoryError
7	NotImplemented
7	NotImplementedError
7	OSError
7	OverflowError
7	PendingDeprecationWarning
7	PermissionError
7	ProcessLookupError
7	RecursionError
7	ReferenceError
7	ResourceWarning
7	RuntimeError
7	RuntimeWarning
7	StopAsyncIteration
7	StopIteration
7	SyntaxError
7	SyntaxWarning
7	SystemError
7	SystemExit
7	TabError
7	TimeoutError
7	True
7	TypeError
7	UnboundLocalError
7	UnicodeDecodeError
7	UnicodeEncodeError
7	UnicodeError
7	UnicodeTranslateError
7	UnicodeWarning
7	UserWarning
7	ValueError
7	Warning
7	ZeroDivisionError
7	__build_class__
7	__debug__
7	__doc__
7	__import__
7	__loader__
7	__name__
7	__package__
7	__spec__
7	abs
7	all
7	any
7	ascii
7	bin
7	bool
7	breakpoint
7	bytearray
7	bytes
7	callable
7	chr
7	classmethod
7	compile
7	complex
7	copyright
7	credits
7	delattr
7	dict
7	dir
7	divmod
7	enumerate
7	eval
7	exec
7	exit
7	filter
7	float
7	format
7	frozenset
7	getattr
7	globals
7	hasattr
7	hash
7	help
7	hex
7	id
7	input
7	int
7	isinstance
7	issubclass
7	iter
7	len
7	license
7	list
7	locals
7	map
7	max
7	memoryview
7	min
7	next
7	object
7	oct
7	open
7	ord
7	pow
7	print
7	property
7	quit
7	range
7	repr
7	reversed
7	round
7	set
7	setattr
7	slice
7	sorted
7	staticmethod
7	str
7	sum
7	super
7	tuple
7	type
7	vars
7	zip
